gamemodes/terrortown/entities/entities/ttt_beacon.lua

```diff
@@ -0,0 +1,222 @@
+---
+-- @class ENT
+-- @section ttt_beacon
+
+if SERVER then
+	AddCSLuaFile()
+end
+
+if CLIENT then
+	ENT.Icon = "vgui/ttt/icon_beacon"
+	ENT.PrintName = "Beacon"
+end
+
+ENT.Type = "anim"
+ENT.Model = Model("models/props_lab/reciever01b.mdl")
+ENT.CanHavePrints = true
+ENT.CanUseKey = true
+
+local beaconDetectionRange = 135
+
+local soundZap = Sound("npc/assassin/ball_zap1.wav")
+local soundBeep = Sound("weapons/c4/cc4_beep1.wav")
+
+---
+-- @realm shared
+function ENT:Initialize()
+	self:SetModel(self.Model)
+
+	if SERVER then
+		self:PhysicsInit(SOLID_VPHYSICS)
+	end
+
+	self:SetMoveType(MOVETYPE_VPHYSICS)
+	self:SetSolid(SOLID_BBOX)
+	self:SetCollisionGroup(COLLISION_GROUP_WEAPON)
+
+	if SERVER then
+		self:SetMaxHealth(100)
+	end
+	self:SetHealth(100)
+
+	if SERVER then
+		self:SetUseType(SIMPLE_USE)
+		self:NextThink(CurTime() + 1)
+	end
+end
+
+---
+-- @param Entity activator
+-- @realm shared
+function ENT:UseOverride(activator)
+	if not IsValid(activator) or self:GetOwner() ~= activator then return end
+
+	local wep = activator:GetWeapon("weapon_ttt_beacon")
+
+	if IsValid(wep) then
+		local pickup = wep:PickupBeacon()
+
+		if not pickup then return end
+	else
+		wep = activator:GiveEquipmentWeapon("weapon_ttt_beacon")
+	end
+
+	self:Remove()
+end
+
+---
+-- @param CTakeDamageInfo dmginfo
+-- @realm shared
+function ENT:OnTakeDamage(dmginfo)
+	self:TakePhysicsDamage(dmginfo)
+	self:SetHealth(self:Health() - dmginfo:GetDamage())
+
+	if self:Health() <= 0 then
+		self:Remove()
+
+		local effect = EffectData()
+		effect:SetOrigin(self:GetPos())
+
+		util.Effect("cball_explode", effect)
+
+		sound.Play(soundZap, self:GetPos())
+
+		if IsValid(self:GetOwner()) then
+			LANG.Msg(self:GetOwner(), "msg_beacon_destroyed", nil, MSG_MSTACK_WARN)
+		end
+	end
+end
+
+---
+-- @realm shared
+function ENT:Think()
+	if SERVER then
+		sound.Play(soundBeep, self:GetPos(), 100, 80)
+	else
+		local entsFound = ents.FindInSphere(self:GetPos(), beaconDetectionRange)
+		local plysFound = {}
+
+		for i = 1, #entsFound do
+			local ent = entsFound[i]
+
+			if not IsValid(ent) or not ent:IsPlayer() then continue end
+
+			plysFound[#plysFound + 1] = ent
+		end
+
+		marks.Remove(self.lastPlysFound or {})
+		marks.Add(plysFound, roles.DETECTIVE.color)
+
+		self.lastPlysFound = plysFound
+	end
+
+	-- only sets the next think on the server
+	self:NextThink(CurTime() + 5)
```

Comment:
Doesn't this mean we effectively render people that just touch the range of the beacon for 5 seconds until they are removed again and not in the radius anymore?

How about adjusting this to be every tick instead and adding some kind of "this frame only"-function to the marks library?

Reply:
no, as the command says this is only sets the next tick for the server. On the client it is updated every tick. It is a bit confusing and it took me a while to understand that part of the code as well. Therefore I added the comment